src/runmode-dpdk.c

```diff
@@ -1102,6 +1248,86 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
     SCReturnInt(0);
 }
 
+static void DeviceRingNameInit(char *name_out, uint8_t name_len, const char *format, uint16_t q_num)
+{
+    char name[RTE_RING_NAMESIZE];
+    char q_num_str[5];
+    char *q_pos;
+
+    if (q_num > 9999)
+        FatalError(SC_ERR_DPDK_CONF, "Number of queues must not exceed 10000 queues per interface");
+    q_pos = strstr(format, DPDK_CONFIG_DEFAULT_QUEUE_NUM_SPECIFIER);
+    snprintf(name, MIN((long)sizeof(name), q_pos - format + 1), "%s", format);
+    snprintf(q_num_str, sizeof(q_num_str), "%" PRIu16, q_num);
+    strlcat(name, q_num_str, sizeof(name));
+    // copy the rest after queue number specifier
+    strlcat(name, q_pos + strlen(DPDK_CONFIG_DEFAULT_QUEUE_NUM_SPECIFIER), sizeof(name));
+    strlcpy(name_out, name, name_len);
+}
+
+static bool DeviceRingNameIsValid(const char *name)
+{
+    uint32_t len = strlen(name);
+    if (len > RTE_RING_NAMESIZE) {
+        SCLogError(SC_ERR_DPDK_CONF, "Ring name (entry \"interface\" %s) cannot be longer than %lu",
+                name, RTE_RING_NAMESIZE);
+        return false;
+    } else if (strstr(name, DPDK_CONFIG_DEFAULT_QUEUE_NUM_SPECIFIER) == NULL) {
+        SCLogError(SC_ERR_DPDK_CONF,
+                "Ring name (entry \"interface\" %s) omits the queue number specifier - \"%s\"",
+                name, DPDK_CONFIG_DEFAULT_QUEUE_NUM_SPECIFIER);
+        return false;
+    }
+    return true;
+}
+
+static int32_t DeviceRingsAttach(DPDKIfaceConfig *iconf)
+{
+    SCEnter();
+    iconf->rx_rings = SCCalloc(iconf->threads, sizeof(struct rte_ring *));
+    if (iconf->rx_rings == NULL) {
+        SCLogError(SC_ERR_DPDK_INIT, "Failed to calloc rx rings");
+        SCReturnInt(-ENOMEM);
+    }
+
+    iconf->tx_rings = SCCalloc(iconf->threads, sizeof(struct rte_ring *));
+    if (iconf->tx_rings == NULL) {
+        SCLogError(SC_ERR_DPDK_INIT, "Failed to calloc tx rings");
+        SCReturnInt(-ENOMEM);
+    }
+
+    if (!DeviceRingNameIsValid(iconf->iface))
```

Comment:
in general I prefer checks like this to be done before we make allocations we then have to unwind